File: src/test/java/no/nav/syfo/service/BehandleFeiledeSoknaderServiceTest.java
package no.nav.syfo.service;

import no.nav.syfo.consumer.repository.InnsendingDAO;
import no.nav.syfo.consumer.ws.AktorConsumer;
import no.nav.syfo.domain.Innsending;
import no.nav.syfo.domain.Soknad;
import no.nav.syfo.domain.dto.Sykepengesoknad;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.time.LocalDate;
import java.util.Collections;

import static no.nav.syfo.domain.dto.Soknadstype.SELVSTENDIGE_OG_FRILANSERE;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class BehandleFeiledeSoknaderServiceTest {

    @Mock
    private AktorConsumer aktorConsumer;

    @Mock
    private InnsendingDAO innsendingDAO;

    @Mock
    private SaksbehandlingsService saksbehandlingsService;

    @InjectMocks
    private BehandleFeiledeSoknaderService behandleFeiledeSoknaderService;

    private Sykepengesoknad sykepengesoknad;
    private Innsending globalInnsending;

    @Before
    public void setup() {
        sykepengesoknad = Sykepengesoknad.builder()
                .id("soknadsId")
                .aktorId("aktorId")
                .soknadstype(SELVSTENDIGE_OG_FRILANSERE)
                .sporsmal(Collections.emptyList())
                .sykmeldingId("sykmeldingsId")
                .fom(LocalDate.of(2018, 9, 7))
                .tom(LocalDate.of(2018, 9, 13))
                .status("SENDT")
                .sendtNav(LocalDate.of(2018, 9, 13).atStartOfDay())
                .opprettet(LocalDate.of(2018, 9, 7).atStartOfDay())
                .build();

        globalInnsending = Innsending.builder()
                .innsendingsId("innsendingsId")
                .ressursId("soknadsId")
                .build();

        when(saksbehandlingsService.opprettSak("innsendingsId", "fnr")).thenReturn("saksId");
        when(saksbehandlingsService.opprettJournalpost(anyString(), any(Soknad.class), anyString())).thenReturn("journalpostId");
        when(aktorConsumer.finnFnr("aktorId")).thenReturn("fnr");
        when(saksbehandlingsService.opprettSoknad(any(), anyString())).thenReturn(Soknad.lagSoknad(sykepengesoknad, "fnr", "Ola Nordmann"));
    }

    @Test
    public void rebehandlerInnsendingSomHarFeiletIAktor() {
        behandleFeiledeSoknaderService.behandleFeiletSoknad(globalInnsending, sykepengesoknad);

        verify(innsendingDAO).oppdaterAktorId("innsendingsId", "aktorId");

        verify(saksbehandlingsService).opprettSak(anyString(), anyString());
        verify(saksbehandlingsService).opprettJournalpost(anyString(), any(Soknad.class), anyString());
        verify(saksbehandlingsService).opprettOppgave(anyString(), anyString(), any(Soknad.class), anyString(), anyString());

        verify(innsendingDAO).settBehandlet("innsendingsId");
        verify(innsendingDAO).fjernFeiletInnsending("innsendingsId");
    }

    @Test
    public void rebehandlerInnsendingSomHarFeiletISak() {
        Innsending innsending = globalInnsending.toBuilder()
                .aktorId("aktorId")
                .build();

        behandleFeiledeSoknaderService.behandleFeiletSoknad(innsending, sykepengesoknad);

        verify(innsendingDAO, never()).oppdaterAktorId("innsendingsId", "aktorId");

        verify(saksbehandlingsService).opprettSak(anyString(), anyString());
        verify(saksbehandlingsService).opprettJournalpost(anyString(), any(Soknad.class), anyString());
        verify(saksbehandlingsService).opprettOppgave(anyString(), anyString(), any(Soknad.class), anyString(), anyString());

        verify(innsendingDAO).settBehandlet("innsendingsId");
        verify(innsendingDAO).fjernFeiletInnsending("innsendingsId");
    }

    @Test
    public void rebehandlerInnsendingSomHarFeiletIJournalPost() {
        Innsending innsending = globalInnsending.toBuilder()
                .aktorId("aktorId")
                .saksId("saksId")
                .build();

        behandleFeiledeSoknaderService.behandleFeiletSoknad(innsending, sykepengesoknad);

        verify(saksbehandlingsService, never()).opprettSak(anyString(), anyString());
        verify(saksbehandlingsService).opprettJournalpost(anyString(), any(Soknad.class), anyString());
        verify(saksbehandlingsService).opprettOppgave(anyString(), anyString(), any(Soknad.class), anyString(), anyString());

        verify(innsendingDAO).settBehandlet("innsendingsId");
        verify(innsendingDAO).fjernFeiletInnsending("innsendingsId");
    }

    @Test
    public void rebehandlerInnsendingSomHarFeiletIOppgave() {
        Innsending innsending = globalInnsending.toBuilder()
                .aktorId("aktorId")
                .saksId("saksId")
                .journalpostId("journalpostId")
                .build();

        behandleFeiledeSoknaderService.behandleFeiletSoknad(innsending, sykepengesoknad);

        verify(saksbehandlingsService, never()).opprettSak(anyString(), anyString());
        verify(saksbehandlingsService, never()).opprettJournalpost(anyString(), any(Soknad.class), anyString());
        verify(saksbehandlingsService).opprettOppgave(anyString(), anyString(), any(Soknad.class), anyString(), anyString());

        verify(innsendingDAO).settBehandlet("innsendingsId");
        verify(innsendingDAO).fjernFeiletInnsending("innsendingsId");
    }


}

File: src/test/java/no/nav/syfo/service/BehandleFeiledeSoknaderServiceTest.kt
package no.nav.syfo.service

import no.nav.syfo.any
import no.nav.syfo.consumer.aktor.AktorConsumer
import no.nav.syfo.consumer.repository.InnsendingDAO
import no.nav.syfo.domain.Innsending
import no.nav.syfo.domain.Soknad
import no.nav.syfo.domain.dto.Soknadstype.SELVSTENDIGE_OG_FRILANSERE
import no.nav.syfo.domain.dto.Sykepengesoknad
import no.nav.syfo.eq
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import java.time.LocalDate

@RunWith(MockitoJUnitRunner::class)
class BehandleFeiledeSoknaderServiceTest {

    @Mock
    private lateinit var aktorConsumer: AktorConsumer

    @Mock
    private lateinit var innsendingDAO: InnsendingDAO

    @Mock
    private lateinit var saksbehandlingsService: SaksbehandlingsService

    @InjectMocks
    private lateinit var behandleFeiledeSoknaderService: BehandleFeiledeSoknaderService

    private lateinit var sykepengesoknad: Sykepengesoknad
    private lateinit var globalInnsending: Innsending

    @Before
    fun setup() {
        sykepengesoknad = Sykepengesoknad(
                id = "soknadsId",
                aktorId = "aktorId",
                soknadstype = SELVSTENDIGE_OG_FRILANSERE,
                sporsmal = emptyList(),
                sykmeldingId = "sykmeldingsId",
                fom = LocalDate.of(2018, 9, 7),
                tom = LocalDate.of(2018, 9, 13),
                status = "SENDT",
                sendtNav = LocalDate.of(2018, 9, 13).atStartOfDay(),
                opprettet = LocalDate.of(2018, 9, 7).atStartOfDay()

        )

        globalInnsending = Innsending("innsendingsId", "soknadsId")

        given(saksbehandlingsService.finnEllerOpprettSak("innsendingsId", "aktorId", sykepengesoknad.fom)).willReturn("saksId")
        given(saksbehandlingsService.opprettJournalpost(anyString(), no.nav.syfo.any(), anyString())).willReturn("journalpostId")
        given(aktorConsumer.finnFnr("aktorId")).willReturn("fnr")
        given(saksbehandlingsService.opprettSoknad(no.nav.syfo.any(), anyString())).willReturn(Soknad.lagSoknad(sykepengesoknad, "fnr", "Ola Nordmann"))
    }

    @Test
    fun rebehandlerInnsendingSomHarFeiletIAktor() {
        behandleFeiledeSoknaderService.behandleFeiletSoknad(globalInnsending, sykepengesoknad)

        verify<InnsendingDAO>(innsendingDAO).oppdaterAktorId("innsendingsId", "aktorId")

        verify<SaksbehandlingsService>(saksbehandlingsService).finnEllerOpprettSak(anyString(), anyString(), any())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettJournalpost(anyString(), no.nav.syfo.any(), anyString())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettOppgave(eq("innsendingsId"), anyString(), anyString(), no.nav.syfo.any(), anyString(), anyString())

        verify<InnsendingDAO>(innsendingDAO).settBehandlet("innsendingsId")
        verify<InnsendingDAO>(innsendingDAO).fjernFeiletInnsending("innsendingsId")
    }

    @Test
    fun rebehandlerInnsendingSomHarFeiletISak() {
        val innsending = globalInnsending.copy(globalInnsending.innsendingsId, globalInnsending.ressursId, "aktorId")

        behandleFeiledeSoknaderService.behandleFeiletSoknad(innsending, sykepengesoknad)

        verify<InnsendingDAO>(innsendingDAO, never()).oppdaterAktorId("innsendingsId", "aktorId")

        verify<SaksbehandlingsService>(saksbehandlingsService).finnEllerOpprettSak(anyString(), anyString(), any())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettJournalpost(anyString(), no.nav.syfo.any(), anyString())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettOppgave(anyString(), anyString(), anyString(), no.nav.syfo.any(), anyString(), anyString())

        verify<InnsendingDAO>(innsendingDAO).settBehandlet("innsendingsId")
        verify<InnsendingDAO>(innsendingDAO).fjernFeiletInnsending("innsendingsId")
    }

    @Test
    fun rebehandlerInnsendingSomHarFeiletIJournalPost() {
        val innsending = globalInnsending.copy(globalInnsending.innsendingsId, globalInnsending.ressursId, "aktorId", "saksId")

        behandleFeiledeSoknaderService.behandleFeiletSoknad(innsending, sykepengesoknad)

        verify<SaksbehandlingsService>(saksbehandlingsService, never()).finnEllerOpprettSak(anyString(), anyString(), any())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettJournalpost(anyString(), no.nav.syfo.any(), anyString())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettOppgave(anyString(), anyString(), anyString(), no.nav.syfo.any(), anyString(), anyString())

        verify<InnsendingDAO>(innsendingDAO).settBehandlet("innsendingsId")
        verify<InnsendingDAO>(innsendingDAO).fjernFeiletInnsending("innsendingsId")
    }

    @Test
    fun rebehandlerInnsendingSomHarFeiletIOppgave() {
        val innsending = globalInnsending.copy(globalInnsending.innsendingsId, globalInnsending.ressursId, "aktorId", "saksId", "journalpostId")

        behandleFeiledeSoknaderService.behandleFeiletSoknad(innsending, sykepengesoknad)

        verify<SaksbehandlingsService>(saksbehandlingsService, never()).finnEllerOpprettSak(anyString(), anyString(), any())
        verify<SaksbehandlingsService>(saksbehandlingsService, never()).opprettJournalpost(anyString(), no.nav.syfo.any(), anyString())
        verify<SaksbehandlingsService>(saksbehandlingsService).opprettOppgave(anyString(), anyString(), anyString(), no.nav.syfo.any(), anyString(), anyString())

        verify<InnsendingDAO>(innsendingDAO).settBehandlet("innsendingsId")
        verify<InnsendingDAO>(innsendingDAO).fjernFeiletInnsending("innsendingsId")
    }
}
